server/encryptionkm/key_manager.go

```diff
@@ -14,41 +14,494 @@
+			}
+			if currentKey.Method != m.method || currentKey.WasExposed ||
+				time.Unix(int64(currentKey.CreationTime), 0).
+					Add(m.dataKeyRotationPeriod).Before(m.helper.now()) {
+				needRotate = true
+			}
+		}
+		if needRotate {
+			rotated := false
+			for attempt := 0; attempt < keyRotationRetryLimit; attempt += 1 {
+				keyID, key, err := encryption.NewDataKey(m.method, uint64(m.helper.now().Unix()))
+				if err != nil {
+					return nil
+				}
+				if keys.Keys[keyID] == nil {
+					keys.Keys[keyID] = key
+					keys.CurrentKeyId = keyID
+					rotated = true
+					log.Info("ready to create or rotate data encryption key", zap.Uint64("keyID", keyID))
+					break
+				}
+				// Duplicated key id. retry.
+			}
+			if !rotated {
+				log.Warn("failed to rotate keys. maximum attempts reached")
+				return errs.ErrEncryptionRotateDataKey.GenWithStack("maximum attempts reached")
+			}
+			needUpdate = true
+		}
+	}
+	if !needUpdate {
+		return nil
+	}
+	// Store updated keys in etcd.
+	err = saveKeys(m.mu.leadership, m.masterKeyMeta, keys)
+	if err != nil {
+		m.helper.eventSaveKeysFailure()
+		log.Error("failed to save keys", zap.Error(err))
+		return err
+	}
+	// Reload keys.
+	_, err = m.loadKeysImpl()
+	return err
+}
+
+func (m *KeyManager) getKeys() *encryptionpb.KeyDictionary {
+	keys := m.keys.Load()
+	if keys == nil {
+		return nil
+	}
+	return keys.(*encryptionpb.KeyDictionary)
 }
 
 // GetCurrentKey get the current encryption key. The key is nil if encryption is not enabled.
 func (m *KeyManager) GetCurrentKey() (keyID uint64, key *encryptionpb.DataKey, err error) {
-	// TODO: Implement
-	return 0, nil, nil
+	keys := m.getKeys()
+	if keys == nil || keys.CurrentKeyId == disableEncryptionKeyID {
+		// Encryption is not enabled.
+		return 0, nil, nil
+	}
+	keyID = keys.CurrentKeyId
+	if keys.Keys == nil {
+		return 0, nil, errs.ErrEncryptionCurrentKeyNotFound.GenWithStack(
+			"empty key list, currentKeyID = %d", keyID)
+	}
+	key = keys.Keys[keyID]
+	if key == nil {
+		// Shouldn't happen, unless key dictionary is corrupted.
+		return 0, nil, errs.ErrEncryptionCurrentKeyNotFound.GenWithStack("currentKeyID = %d", keyID)
+	}
+	return keyID, key, nil
 }
 
-// GetKey get the encryption key with the specific key id.
-func (m *KeyManager) GetKey(keyID uint64) (key *encryptionpb.DataKey, err error) {
-	// TODO: Implement
-	return nil, nil
+// getKeyLocal gets specific encryption key by key id, from local cache.
+func (m *KeyManager) getKeyLocal(keyID uint64) *encryptionpb.DataKey {
+	keys := m.getKeys()
+	if keys == nil || keys.Keys == nil {
+		return nil
+	}
+	return keys.Keys[keyID]
+}
+
+// GetKey gets specific encryption key by key id.
+func (m *KeyManager) GetKey(keyID uint64) (*encryptionpb.DataKey, error) {
+	key := m.getKeyLocal(keyID)
+	if key != nil {
+		return key, nil
+	}
+	// Key not found in memory.
+	// The key could be generated by another PD node, which shouldn't happen normally.
```

Comment:
Does it need a warning log here?

Reply:
I'm hesitate to add logging on a frequently access code path.